### PR TITLE
Add ConvertDenseToFsaVec()

### PR DESCRIPTION
**Pull request by danpovey**

### CODE REVIEW

**Comment by qindazhu** [comment]
Seem you did not rebase on the latest master branch. There are conflicts.

**Comment by danpovey** [comment]
Guys, I merged but it needs to be tested.  Does anyone  have time to do that?
This function is to be used in testing the composition code; I intend to use it to convert to regular FSA so we can compare with CPU-based composition.  Once there is a test of the Compose() code I can debug it myself.

**Comment by qindazhu** [comment]
I will do this tomorrow as Fangjun may be busy with RandomFsa stuffs.

**Comment by csukuangfj** [comment]
I am adding test code for it. 

RandomFsa is not that urgent and I'll implement it on Saturday.

**Comment by qindazhu** [comment]
OK, thanks!

File: k2/csrc/fsa_utils.cu [hunk 1]
```diff
@@ -925,4 +925,77 @@ Ragged<int32_t> GetEnteringArcIndexBatches(FsaVec &fsas,
...
+      symbol_offset = -1;
+      if (s > 0) return;  // we just need the arc with -1.
+    } else {
+      symbol_offset = 0;
+    }
+    // the "+ 1" is because index 0 in `scores` is for the final-symbol -1,
+    // then 0, 1, etc.
+    int32_t symbol_index_in_scores = s + symbol_offset + 1;
+    arcs_data[arc_idx012] = Arc(state_idx1, state_idx1 + 1, s + symbol_offset,
+                                FloatAsInt(scores_acc(src_state_idx01,
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:983]: Do we need `FloatAsInt` here?

`scores_acc` is already of type float and the constructor of `Arc` also needs a float
score.

**Comment by danpovey** [comment]
Oh, you're right.

File: k2/csrc/fsa_utils.cu [hunk 2]
```diff
@@ -925,4 +925,77 @@ Ragged<int32_t> GetEnteringArcIndexBatches(FsaVec &fsas,
...
+
+    // arc_idx0x is the 1st arc-index of the FSA we are creating.. each source
+    // state has `num_symbols` arcs leaving it except the last one of each FSA,
+    // which has 1 arc leaving it (to the final-state).
+    int32_t arc_idx0xx = (src_state_idx0x * num_symbols) -
+      fsa_idx0 * (num_symbols - 1),
+      arc_idx01x = arc_idx0xx + (state_idx1 * num_symbols),
+      arc_idx012 = arc_idx01x + s;
+    int32_t symbol_offset;
+    if (state_idx1 + 1 < src_num_states1) {
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:973]: should this be `state_idx1 + 1 == src_num_states1` ?

Comment (danpovey) [reply k2/csrc/fsa_utils.cu:973]: you're right.

**Comment by csukuangfj** [comment]
Will fix them.

**Pull request merged**